R/CreatePackageReport.R

```diff
@@ -8,122 +265,70 @@
                                 ) {
-    # Input checks
-    assertthat::assert_that(
-        assertthat::is.string(pkg_name)
-        , pkg_name != ""
-        , is.list(pkg_reporters)
-        , is.null(pkg_path) || assertthat::is.readable(pkg_path)
-        , assertthat::is.string(report_path)
-        , report_path != ""
-    )
 
-    # Confirm that the report_path looks correct
-    if (! identical(tolower(tools::file_ext(report_path)), "html")){
-        log_fatal(sprintf("report_path must be a .html file path. You gave '%s'.", report_path))
-    }
+    # pkg_name, pkg_path, report_path validated by PackageReport
 
-    # Confirm that all reporters are actually reporters
-    checks <- sapply(pkg_reporters, function(x){methods::is(x, "AbstractPackageReporter")})
-    if (!all(checks)){
-        msg <- paste0("At least one of the reporters in the pkg_reporters parameter ",
-                      "is not a PackageReporter. See ?AbstractPackageReporter for details.")
-        log_fatal(msg)
+    ## pkg_reporters input checks ##
+    assertthat::assert_that(
+        is.list(pkg_reporters)
+    )
+    # Check if generators were passed in by accident
+    if (any(vapply(pkg_reporters, FUN = R6::is.R6Class, FUN.VALUE = logical(1)))) {
+        log_fatal(paste(
+            "At least one of pkg_reporters is an R6 class generator. This"
+            , "function expects initialized reporter objects."
+        ))
     }
+    # Confirm that all reporters are actually valid initialized reporters
+    assertthat::assert_that(
+        all(vapply(pkg_reporters
+                   , FUN = .is.PackageReporter
+                   , FUN.VALUE = logical(1)
+        ))
+        , msg = "All members of pkg_reporters must be initialized package reporters."
+    )
+    # Confirm that all reporters are recognized by PackageReport
+    lapply(
+        X = pkg_reporters
+        , FUN = function(reporter) {
+            assertthat::assert_that(
+                class(reporter)[1] %in% names(PackageReport$active)
+            )
+        }
+    )
 
     log_info(paste0("Creating package report for package "
                     , pkg_name
                     , " with reporters:\n\n"
                     , paste(unlist(lapply(pkg_reporters, function(x) class(x)[1]))
                             , collapse = "\n")))
 
-    builtReporters <- .BuildPackageReporters(
-      pkg_name
-      , pkg_reporters
-      , pkg_path
+    ## Create PackageReport object ##
+    createdReport <- PackageReport$new(
+        pkg_name = pkg_name,
+        pkg_path = pkg_path,
+        report_path = report_path
     )
-
-    .RenderPackageReport(
-      report_path = report_path
-      , pkg_reporters = builtReporters
-      , pkg_name = pkg_name
+    lapply(
+        X = pkg_reporters
+        , FUN = function(reporter) {
+            class <- class(reporter)[1]
+            createdReport[[class]] <- reporter
```

Comment:
@jayqi can you explain a little more what's going on here?  Is this syntax to refer to the active methods that are built in `PackageReport`?  What are your thoughts on an `AddReporter` method for `PackageReport`?

Reply:
Yeah, this is how active bindings work. So in general you can do `myObject$myBinding <- 'value'` or whatever. 

That's a good idea about adding a method. I agree that the current method is somewhat clunky for interactive use.